File: python/relink_py/tests/test_adaptive_bitrate.py
"""AdaptiveBitrateController tests -- pure arithmetic, no sockets, no
codec. Mirrors tests/test_adaptive_bitrate.cpp."""
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from relink.adaptive_bitrate import AdaptiveBitrateController

failures = 0


def check(cond, desc):
    global failures
    if cond:
        print(f"ok: {desc}")
    else:
        print(f"FAIL: {desc}", file=sys.stderr)
        failures += 1


# --- motion-based: low motion -> high quality, high motion -> low quality ---
c = AdaptiveBitrateController(quality_min=20, quality_max=80, motion_ceiling=30.0, smoothing=1.0)
check(c.next_quality(0.0, now=0.0) == 80, "zero motion maps to quality_max")
check(c.next_quality(60.0, now=0.1) == 20, "motion at/above ceiling maps to quality_min")

# --- smoothing: quality should move gradually, not jump ---
c2 = AdaptiveBitrateController(quality_min=20, quality_max=80, motion_ceiling=30.0, smoothing=0.3)
prev = c2.current_quality
check(prev == 80, "starts at quality_max")
t = 0.0
monotonic = True
for _ in range(10):
    q = c2.next_quality(60.0, now=t)
    if q > prev:
        monotonic = False
    prev = q
    t += 0.1
check(monotonic, "smoothed quality trends monotonically down toward quality_min")
check(20 <= prev < 80, "smoothed quality settles strictly below quality_max but within range")

# --- bitrate ceiling: exceeding target bitrate pulls quality down even
# with zero motion ---
c3 = AdaptiveBitrateController(quality_min=20, quality_max=80, motion_ceiling=30.0,
                                target_bitrate_bps=1_000_000, window_sec=1.0, smoothing=1.0)
t = 0.0
q = 80
for _ in range(5):
    q = c3.next_quality(0.0, now=t)
    # 200KB/frame * 5/sec = 8Mbps, way over the 1Mbps target
    c3.record_sent(200_000, now=t)
    t += 0.2
check(q < 80, "quality pulled down from the motion-only baseline of 80 under bitrate pressure")
check(q >= 20, "quality never drops below quality_min")

# --- invalid config is rejected loudly ---
threw = False
try:
    AdaptiveBitrateController(quality_min=90, quality_max=10)
except ValueError:
    threw = True
check(threw, "quality_min > quality_max is rejected")

threw = False
try:
    AdaptiveBitrateController(smoothing=0.0)
except ValueError:
    threw = True
check(threw, "smoothing of 0 is rejected (would never move)")

print()
if failures == 0:
    print("ALL PASS")
    sys.exit(0)
else:
    print(f"{failures} FAILURE(S)")
    sys.exit(1)

File: cpp/tests/test_adaptive_bitrate.cpp
// Step: AdaptiveBitrateController unit tests. Pure arithmetic, no
// sockets, no codec -- mirrors relink_py/tests/test_adaptive_bitrate.py.

#include "relink/adaptive_bitrate.hpp"
#include <cstdio>

using namespace relink;

static int g_failures = 0;
#define CHECK(cond) do { \
    if (!(cond)) { std::fprintf(stderr, "FAIL: %s (%s:%d)\n", #cond, __FILE__, __LINE__); ++g_failures; } \
    else { std::printf("ok: %s\n", #cond); } \
} while (0)

int main() {
    // --- motion-based: low motion -> high quality, high motion -> low quality ---
    {
        AdaptiveBitrateController::Config cfg;
        cfg.quality_min = 20;
        cfg.quality_max = 80;
        cfg.motion_ceiling = 30.0;
        cfg.smoothing = 1.0; // no smoothing -- jump straight to target for this check
        AdaptiveBitrateController c(cfg);
        CHECK(c.next_quality(0.0, 0.0) == 80);
        CHECK(c.next_quality(60.0, 0.1) == 20); // >= ceiling clamps to quality_min
    }

    // --- smoothing: quality should move gradually, not jump ---
    {
        AdaptiveBitrateController::Config cfg;
        cfg.quality_min = 20;
        cfg.quality_max = 80;
        cfg.motion_ceiling = 30.0;
        cfg.smoothing = 0.3;
        AdaptiveBitrateController c(cfg);
        int prev = c.current_quality();
        CHECK(prev == 80);
        double t = 0.0;
        for (int i = 0; i < 10; ++i) {
            int q = c.next_quality(60.0, t);
            CHECK(q <= prev); // monotonically trending down toward quality_min
            prev = q;
            t += 0.1;
        }
        CHECK(prev < 80 && prev >= 20);
    }

    // --- bitrate ceiling: exceeding target bitrate pulls quality down
    // even with zero motion ---
    {
        AdaptiveBitrateController::Config cfg;
        cfg.quality_min = 20;
        cfg.quality_max = 80;
        cfg.motion_ceiling = 30.0;
        cfg.target_bitrate_bps = 1'000'000.0;
        cfg.window_sec = 1.0;
        cfg.smoothing = 1.0;
        AdaptiveBitrateController c(cfg);
        double t = 0.0;
        int q = 80;
        for (int i = 0; i < 5; ++i) {
            q = c.next_quality(0.0, t);
            // 200KB/frame * 5/sec = 8Mbps, way over the 1Mbps target
            c.record_sent(200000, t);
            t += 0.2;
        }
        CHECK(q < 80); // pulled down from the motion-only baseline of 80
        CHECK(q >= 20);
    }

    // --- invalid config is rejected loudly ---
    {
        bool threw = false;
        try {
            AdaptiveBitrateController::Config cfg;
            cfg.quality_min = 90;
            cfg.quality_max = 10; // min > max
            AdaptiveBitrateController c(cfg);
        } catch (const std::invalid_argument&) {
            threw = true;
        }
        CHECK(threw);
    }

    if (g_failures == 0) { std::printf("\nALL PASS\n"); return 0; }
    std::printf("\n%d FAILURE(S)\n", g_failures);
    return 1;
}
